lib/elastic_apm/transport/base.rb

```diff
@@ -8,90 +8,156 @@
     class Base
       include Logging
 
+      WATCHER_EXECUTION_INTERVAL = 5
+      WATCHER_TIMEOUT_INTERVAL = 4
+      WORKER_JOIN_TIMEOUT = 5
+
       def initialize(config)
         @config = config
         @queue = SizedQueue.new(config.api_buffer_size)
-        @pool = Concurrent::FixedThreadPool.new(config.pool_size)
-        @workers = []
 
         @serializers = Serializers.new(config)
         @filters = Filters.new(config)
+
+        @stopped = Concurrent::AtomicBoolean.new
+        @workers = Array.new(config.pool_size)
+
+        @watcher_mutex = Mutex.new
+        @worker_mutex = Mutex.new
       end
 
-      attr_reader :config, :queue, :workers, :filters
+      attr_reader :config, :queue, :filters, :workers, :watcher, :stopped
 
       def start
+        debug '%s: Starting Transport', pid_str
+
+        ensure_watcher_running
+        ensure_worker_count
       end
 
       def stop
+        debug '%s: Stopping Transport', pid_str
+
+        @stopped.make_true
+
+        stop_watcher
         stop_workers
       end
 
+      # rubocop:disable Metrics/MethodLength, Metrics/LineLength
       def submit(resource)
+        if @stopped.true?
+          warn '%s: Transport stopping, no new events accepted', pid_str
+          return false
+        end
+
+        ensure_watcher_running
         queue.push(resource, true)
 
-        ensure_worker_count
+        true
       rescue ThreadError
-        warn 'Queue is full (%i items), skipping…', config.api_buffer_size
+        warn '%s: Queue is full (%i items), skipping…', pid_str, config.api_buffer_size
+        nil
+      rescue Exception => e
+        error '%s: Failed adding to the transport queue: %p', pid_str, e.inspect
         nil
       end
+      # rubocop:enable Metrics/MethodLength, Metrics/LineLength
 
       def add_filter(key, callback)
         @filters.add(key, callback)
       end
 
       private
 
+      def pid_str
+        format('[PID:%s]', Process.pid)
+      end
+
+      def ensure_watcher_running
+        # pid has changed == we've forked
+        return if @pid == Process.pid
+
+        @watcher_mutex.synchronize do
+          return if @pid == Process.pid
+          @pid = Process.pid
+
+          @watcher = Concurrent::TimerTask.execute(
+            execution_interval: WATCHER_EXECUTION_INTERVAL,
+            timeout_interval: WATCHER_TIMEOUT_INTERVAL
+          ) { ensure_worker_count }
+        end
+      end
+
       def ensure_worker_count
-        missing = config.pool_size - @workers.length
-        return unless missing > 0
+        return if all_workers_alive?
+
+        @worker_mutex.synchronize do
+          return if stopped.true?
+
+          @workers.map! do |thread|
+            next thread if thread&.alive?
 
-        info 'Booting %i workers', missing
-        missing.times { boot_worker }
+            boot_worker
+          end
+        end
+      end
+
+      def all_workers_alive?
+        !!workers.all? { |t| t&.alive? }
```

Comment:
I thought about it, but don't think it is necessary. This check is run every n seconds from the watcher thread, and is only a safety measure in case something unexpectedly killed the workers. So worse case, a dead worker is caught a bit later than possible when not synchronized. I think this is better than adding sync overhead within every `submit`.

Reply:
> I think this is better than adding sync overhead within every submit.

Do you mean because the submit calls `ensure_watcher_running`? If so, it won't be _every_ call to submit that's affected, only the ones straight after a fork, due to a pid mismatch, right?

My concern was about concurrent modification due to `@workers.clear` in stop_workers. I guess that's not a problem in MRI, but what about JRuby?

Reply:
You are right, that only concerns the calls after the fork. 
The `workers.clear` could be a problem here. How about protecting `return if all_workers_alive?` in the `ensure_worker_count`?

Reply:
> The workers.clear could be a problem here. How about protecting return if all_workers_alive? in the ensure_worker_count?

That's what I meant by:

> If so, maybe make a note of that requirement here, and move the call to inside the existing synchronize block in ensure_worker_count?

Did you mean this, or something else?